replica/table.cc

```diff
@@ -609,17 +640,11 @@ void table::update_stats_for_new_sstable(const sstables::shared_sstable& sst) no
 future<>
 table::do_add_sstable_and_update_cache(sstables::shared_sstable sst, sstables::offstrategy offstrategy) {
     auto permit = co_await seastar::get_units(_sstable_set_mutation_sem, 1);
-    co_return co_await get_row_cache().invalidate(row_cache::external_updater([this, sst, offstrategy] () noexcept {
-        // FIXME: this is not really noexcept, but we need to provide strong exception guarantees.
-        // atomically load all opened sstables into column family.
-        compaction_group& cg = compaction_group_for_sstable(sst);
-        if (!offstrategy) {
-            add_sstable(cg, sst);
-        } else {
-            add_maintenance_sstable(cg, sst);
-        }
-        update_stats_for_new_sstable(sst);
-    }), dht::partition_range::make({sst->get_first_decorated_key(), true}, {sst->get_last_decorated_key(), true}));
+    auto& cg = compaction_group_for_sstable(sst);
+    auto sstables = std::vector<sstables::shared_sstable>({sst});
+    auto adder = row_cache::external_updater(std::make_unique<table_sstables_adder>(*this, cg, std::move(sstables), is_main(!bool(offstrategy))));
+    co_return co_await get_row_cache().invalidate(std::move(adder),
+            dht::partition_range::make({sst->get_first_decorated_key(), true}, {sst->get_last_decorated_key(), true}));
```

Comment:
Note this can throw

Reply:
True, but that would happen before calling "invalidate" so it's orthogonal to the updater begin exception safe.
Why is it a problem?